google/cloud/storage/internal/signed_url_requests.h

```diff
@@ -27,25 +27,69 @@ namespace cloud {
 namespace storage {
 inline namespace STORAGE_CLIENT_NS {
 namespace internal {
+/// The common data for SignUrlRequests.
+class SignUrlRequestCommon {
+ public:
+  SignUrlRequestCommon() = default;
+  SignUrlRequestCommon(std::string verb, std::string bucket_name,
+                       std::string object_name)
+      : verb_(std::move(verb)),
+        bucket_name_(std::move(bucket_name)),
+        object_name_(std::move(object_name)) {}
+
+  std::string const& verb() const { return verb_; }
+  std::string const& bucket_name() const { return bucket_name_; }
+  std::string const& object_name() const { return object_name_; }
+  std::string const& sub_resource() const { return sub_resource_; }
+
+ protected:
+  void SetOption(SubResourceOption const& o) {
+    if (!o.has_value()) {
+      return;
+    }
+    sub_resource_ = o.value();
+  }
+
+  void SetOption(AddExtensionHeaderOption const& o);
+
+  void SetOption(AddQueryParameterOption const& o) {
+    if (!o.has_value()) {
+      return;
+    }
+    query_parameters_.insert(o.value());
+  }
+
+  std::string verb_;
+  std::string bucket_name_;
+  std::string object_name_;
+  std::string sub_resource_;
+  std::map<std::string, std::string> extension_headers_;
+  std::multimap<std::string, std::string> query_parameters_;
+};
+
 /**
  * Requests the Google Cloud Storage service account for a project.
  */
-class SignUrlRequest {
+class SignUrlRequest : private SignUrlRequestCommon {
```

Comment:
Q. Do we have any style preference for composition over private inheritance?

Reply:
Not that I know of.  This seems like a decent use for private inheritance, but obviously we can change it (at a cost of more code I believe).

Reply:
FWIW: https://google.github.io/styleguide/cppguide.html#Inheritance

TL;DR: Inheritance should be public. Prefer composition over inheritance.

Reply:
Private inheritance **is** composition in my book, but meh.  Should I change this?

Reply:
I personally won't request a change in this PR. But perhaps we should talk about this a bit more separately so that as a team we can come to some agreement one way or the other on the pattern to use going forward.

Also, I agree that private inheritance is very similar to composition, but I don't view them as equivalent. All inheritance, including private, creates a tighter coupling between classes, for one example, because the subclass has access to the parent's protected members. Composition provides a looser coupling between classes because each class can only access the public interface of the other.... and maybe I've just been swayed too much by Scott Meyers's advice on private inheritance :-)

Reply:
Made the change, not worth worrying about it.